api/cluster/controller.go

```diff
@@ -142,75 +167,93 @@ func newController(kfservingClient kservev1beta1client.ServingV1beta1Interface,
+		maxCPU, _ := c.deploymentConfig.MaxCPU.AsInt64()
 		if cpuRequest > maxCPU {
 			log.Errorf("insufficient available cpu resource to fulfil user request of %d", cpuRequest)
 			return nil, ErrInsufficientCPU
 		}
 		memRequest, _ := modelService.ResourceRequest.MemoryRequest.AsInt64()
-		maxMem, _ := k.deploymentConfig.MaxMemory.AsInt64()
+		maxMem, _ := c.deploymentConfig.MaxMemory.AsInt64()
 		if memRequest > maxMem {
 			log.Errorf("insufficient available memory resource to fulfil user request of %d", memRequest)
 			return nil, ErrInsufficientMem
 		}
 	}
 
-	_, err := k.namespaceCreator.CreateNamespace(ctx, modelService.Namespace)
+	_, err := c.namespaceCreator.CreateNamespace(ctx, modelService.Namespace)
 	if err != nil {
 		log.Errorf("unable to create namespace %s %v", modelService.Namespace, err)
 		return nil, ErrUnableToCreateNamespace
 	}
 
 	isvcName := modelService.Name
-	s, err := k.servingClient.InferenceServices(modelService.Namespace).Get(isvcName, metav1.GetOptions{})
+	s, err := c.kserveClient.InferenceServices(modelService.Namespace).Get(isvcName, metav1.GetOptions{})
 	if err != nil {
 		if !kerrors.IsNotFound(err) {
 			log.Errorf("unable to check inference service %s %v", isvcName, err)
 			return nil, ErrUnableToGetInferenceServiceStatus
 		}
 
 		// create new resource
-		spec, err := k.kfServingResourceTemplater.CreateInferenceServiceSpec(modelService, k.deploymentConfig)
+		spec, err := c.kfServingResourceTemplater.CreateInferenceServiceSpec(modelService, c.deploymentConfig)
 		if err != nil {
 			log.Errorf("unable to create inference service spec %s %v", isvcName, err)
 			return nil, ErrUnableToCreateInferenceService
 		}
 
-		s, err = k.servingClient.InferenceServices(modelService.Namespace).Create(spec)
+		s, err = c.kserveClient.InferenceServices(modelService.Namespace).Create(spec)
 		if err != nil {
 			log.Errorf("unable to create inference service %s %v", isvcName, err)
 			return nil, ErrUnableToCreateInferenceService
 		}
 	} else {
-		patchedSpec, err := k.kfServingResourceTemplater.PatchInferenceServiceSpec(s, modelService, k.deploymentConfig)
+		// Get current scale of the existing deployment
+		deploymentScale := resource.DeploymentScale{}
+		if modelService.DeploymentMode == deployment.ServerlessDeploymentMode ||
+			modelService.DeploymentMode == deployment.EmptyDeploymentMode {
+			if predictorScale, err := c.GetCurrentDeploymentScale(
+				ctx, modelService.Namespace,
+				s.Status.Components[kservev1beta1.PredictorComponent].LatestCreatedRevision); err != nil {
+				deploymentScale.Predictor = &predictorScale
+			}
+			if _, ok := s.Status.Components[kservev1beta1.TransformerComponent]; ok {
+				if transformerScale, err := c.GetCurrentDeploymentScale(
+					ctx, modelService.Namespace,
+					s.Status.Components[kservev1beta1.TransformerComponent].LatestCreatedRevision); err != nil {
+					deploymentScale.Transformer = &transformerScale
+				}
+			}
+		}
```

Comment:
Have you considered to merge getting prediction and transformer's deployment scale in one function?

So, in .Deploy() function, it would be something like this:

```
// Get current scale of the existing deployment
		deploymentScale := resource.DeploymentScale{}
		if modelService.DeploymentMode == deployment.ServerlessDeploymentMode ||
			modelService.DeploymentMode == deployment.EmptyDeploymentMode {
			deploymentScale, err = c.getCurrentDeploymentScale(ctx, modelService.Namespace,
				s.Status.Components)
			...
		}
```

Then, the function c.getCurrentDeploymentScale() will be the one that have the logic on whether the transformer component exists or not. I prefer to have .Deploy() function to have higher level of asbtraction.